Feat: (#35) RefreshToken 재발급 API를 개발한다.

HoyeongJeon opened the pull request:
## ✅ PR 유형
어떤 변경 사항이 있었나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [x] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [x] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

---

## 📝 작업 내용
- RefreshToken이 Redis에 존재하는 경우, AccessToken 재발급 로직을 구현했습니다. 
- 토큰 재발급 용 RotateTokenService를 구현했습니다. RefreshTokenService 내에 구현을 할까 고민했지만, RefreshToken의 여부를 확인하고 AccessToken을 발급하는 로직이기에 따로 분리했습니다. 
- 토큰 재발급 Controller를 작성했습니다.
- 토큰 재발급 응답을 구현했습니다.(RotateTokenResponse)

![image](https://github.com/user-attachments/assets/d359f6ee-293e-44a2-ac7c-be8659f9d3d3)

---

## ✏️ 관련 이슈

- Related to : #35 

---

## 🎸 기타 사항 or 추가 코멘트
유저에게 토큰을 전달할 때의 방식을 고민하고 있습니다. 
현재 우리 방식은 유저에게 AccessToken과 RefreshToken을 둘 다 반환합니다. 
RefreshToken은 쿠키에 넣어주는 방식에 대해선 어떻게 생각하는지 궁금합니다~!
RefreshToken을 쿠키에 담을 경우
- httpOnly 설정으로 브라우저에서 js를 통한 접근을 방지할 수 있고,
- secure 옵션을 통해 HTTPS 통신에서만 쿠키를 사용할 수 있도록 강제해 더 안전하게 RefreshToken을 관리할 수 있을거 같습니다 ㅎㅎ

HoyeongJeon committed:
feat: (#35) Redis 토큰 삭제 메서드를 구현한다.

HoyeongJeon committed:
feat: (#35) 토큰 에러를 추가한다.

HoyeongJeon committed:
feat: (#35) RefreshTokenService의 Test를 작성한다.

HoyeongJeon committed:
feat: (#35) RotateToken의 응답을 생성하고, RotateTokenService를 구현한다.

HoyeongJeon committed:
feat: (#35) RotateTokenController를 생성한다.

HoyeongJeon committed:
feat: (#35) 레디스에 memberId에 해당하는 refreshToken이 존재하지 않는 경우 exception을 반환한다.

anxi01 commented on src/main/java/spring/backend/auth/application/RefreshTokenService.java:
```diff
@@ -38,7 +41,24 @@ public String saveRefreshToken(Member member) {
+            return refreshToken;
+        } catch (RedisConnectionException e) {
+            throw GlobalErrorCode.REDIS_CONNECTION_ERROR.toException();
+        } catch (DomainException e) {
```
DomainException은 위에 AuthenticationErrorCode에서 처리하며,
`NOT_EXIST_REFRESH_TOKEN` 이 에러가 아닌 다른 에러는 잡히지 않는 것으로 보입니다.

HoyeongJeon replied:
`
catch(DomainException e)
`
이 부분을 삭제하면 
```Java
  try {
            String refreshToken =  refreshTokenRepository.findByMemberId(memberId);
            if (refreshToken == null || refreshToken.isEmpty()) {
                log.error("리프레시 토큰이 저장소에 존재하지 않습니다.");
                throw AuthenticationErrorCode.NOT_EXIST_REFRESH_TOKEN.toException();
            }
            return refreshToken;
```
위의 throw에서 던지는 값이 
Catch(Exception e)에 잡혀 INTERNAL_ERROR로 잡힙니다. 

![image](https://github.com/user-attachments/assets/e9b70e7a-0796-4add-9cb1-dd95c55d3927)

이 부분에 대한 리뷰가 맞을까요?

anxi01 replied:
아 제가 위 리뷰를 올리면서 try-catch가 없을 경우를 생각했었네요..
위의 경우는 catch(DomainException e)를 제거하면 애러가 잡히지 않는게 맞습니다!
혼동일으켜 죄송합니다ㅜㅜ

HoyeongJeon replied:
괜차나요~! 저도 다시 한번 확인할 수 있어서 좋았습니다!!

anxi01 commented on src/main/java/spring/backend/auth/application/RefreshTokenService.java:
```diff
@@ -38,7 +41,24 @@ public String saveRefreshToken(Member member) {
+                throw AuthenticationErrorCode.NOT_EXIST_REFRESH_TOKEN.toException();
+            }
+            return refreshToken;
+        } catch (RedisConnectionException e) {
```
추가로, 지금 서비스 로직에서 외부 모듈에 대한 Exception이 발생하는데
이 Exception을 Repository에서 처리하면 서비스 로직에서는 순수 비즈니스 로직으로 수행할 수 있을 거라 생각합니다

anxi01 replied:
Exception 처리를 Repository에 위임하는게 어떨까요?

anxi01 replied:
```AuthenticationErrorCode.NOT_EXIST_REFRESH_TOKEN.toException()```
이건 도메인 에러이니 제외하고요!

HoyeongJeon replied:
> `AuthenticationErrorCode.NOT_EXIST_REFRESH_TOKEN.toException()` 이건 도메인 에러이니 제외하고요!

이 부분을 제외하라게 어떤 의미인지 잘 이해를 못했어요 ㅜ

anxi01 replied:
```
try {
            String refreshToken =  refreshTokenRepository.findByMemberId(memberId);
            if (refreshToken == null || refreshToken.isEmpty()) {
                log.error("리프레시 토큰이 저장소에 존재하지 않습니다.");
                throw AuthenticationErrorCode.NOT_EXIST_REFRESH_TOKEN.toException();
            }
            return refreshToken;
        } catch (RedisConnectionException e) {
            throw GlobalErrorCode.REDIS_CONNECTION_ERROR.toException();
        } catch (DomainException e) {
            throw e;
        } catch (Exception e) {
            throw GlobalErrorCode.INTERNAL_ERROR.toException();
        }
```
여기서 에러가 발생하는 지점이 두 곳이 있는데
1. refreshToken이 없을 때 -> AuthenticationErrorCode 발생
2. findByMemberId()에서 내부 에러 발생 -> RedisConnectionException or Exception 발생

1번의 경우, memberId가 잘못되었을 경우 혹은 토큰이 없을 경우 발생하는 에러인데 이 경우엔 해당 서비스에서 에러를 처리하면 좋을 거 같고 `throw AuthenticationErrorCode.NOT_EXIST_REFRESH_TOKEN.toException();`

2번의 경우, 내부 에러가 발생하는 것은 결국 외부 기술에서의 에러이기 때문에 Repository에서 try-catch를 사용해서 에러를 잡는게 어떨까 했습니다!

anxi01 commented on src/main/java/spring/backend/auth/presentation/RotateTokenController.java:
```diff
@@ -0,0 +1,25 @@
+import spring.backend.core.presentation.RestResponse;
+
+@RestController
+@RequestMapping("/v1/token/rotate")
```
`@RequestMapping` 에는 v1만 존재하고 이외 값들은 `@GetMapping` 에서 처리하는게 좋을 거 같습니다.
추후 다른 API가 해당 클래스에 추가될 경우 RequestMapping에서 URI를 다 처리하기 때문에 `v1/token/이외 값`이 들어올 경우 수정이 필요합니다.

anxi01 replied:
@RequestMapping에 URI를 넣을 수도 있고, @GetMapping에도 넣을 수 있어요!
혹시나 해서 알려드립니다~!

anxi01 commented on src/main/java/spring/backend/auth/presentation/RotateTokenController.java:
```diff
@@ -0,0 +1,25 @@
+
+    @GetMapping
+    public ResponseEntity<RestResponse<RotateTokenResponse>> rotateAccessToken(
+            @CookieValue(name = "refreshToken", required = false) String refreshToken
```
q) `required = false` 일 경우, rotateAccessToken 메서드에서 무조건 에러가 발생하는데 required를 false로 한 이유가 있을까요?

HoyeongJeon replied:
> q) `required = false` 일 경우, rotateAccessToken 메서드에서 무조건 에러가 발생하는데 required를 false로 한 이유가 있을까요?

rotateAccessToken에서 무조건 에러가 발생하나요?? (이 부분은 저는 에러가 안나서 어떤 에러를 말하는건지 모르겠어요..)

`required = false` 의 경우, 쿠키에 CookieValue의 name (코드에선 "refreshToken")에 해당하는 값이 없어도 에러를 발생시키지 않으려고 넣었습니다.

`required = true`로 지정하면 요청에 쿠키가 없다면 에러를 발생시킵니다. 

![image](https://github.com/user-attachments/assets/fff35b01-2923-4404-8f79-40cfab67a74b)

이 부분을 잡기 위해서 Controller 안에 MissingCookieValueException 처리를 해주면 됩니다.

```Java
public class RotateTokenController {
    private final RotateTokenService rotateTokenService;

    @GetMapping
    public ResponseEntity<RestResponse<RotateTokenResponse>> rotateAccessToken(
            @CookieValue(name = "refreshToken", required = true) String refreshToken
    ) {
        return ResponseEntity.ok(new RestResponse<>(rotateTokenService.rotateAccessToken(refreshToken)));
    }

    @ExceptionHandler(MissingRequestCookieException.class)
    public ResponseEntity<String> handleMissingCookieValueException(MissingRequestCookieException ex) {
        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body("리프레시 토큰이 누락되었습니다.");
    }
}
```
제가 느끼기엔 이렇게 되면 Controller의 역할을 벗어난다 생각했습니다. (다시 찾아보니 HTTP 요청을 받아 파라미터를 검증하는 것도 Controller의 역할이네요. 코드를 작성하던 당시엔 적절하지 않다고 생각했습니다. ㅎㅎ;)

지금 코드를 보니까 에러 처리를 안했는데, false를 통해 refreshToken이 쿠키에 없어도 NULL 값으로 받고 , Service 로직으로 처리하는 것이 더 깔끔하다 생각해서 false로 처리했습니다. 
```Java
public class RotateTokenService {
    private final MemberService memberService;
    private final JwtService jwtService;
    private final RefreshTokenService refreshTokenService;

    public RotateTokenResponse rotateAccessToken(String refreshToken) {
        UUID memberId = extractMemberIdFromRefreshToken(refreshToken);
        validateRefreshToken(memberId, refreshToken);
        return new RotateTokenResponse(jwtService.provideAccessToken(memberService.findByMemberId(memberId)));
    }

    private UUID extractMemberIdFromRefreshToken(String refreshToken) {
        if(refreshToken == null) {
            log.error("리프레시 토큰이 존재하지 않습니다.");
            throw AuthenticationErrorCode.MISSING_COOKIE_VALUE.toException();
        }
        return UUID.fromString(jwtService.getPayload(refreshToken).get("memberId", String.class));
    }
}
```
(위 코드는 현재 구현하지 않았습니다. 추가로 구현해야 할 부분입니다.)

이렇게 쿠키값이 없는 경우 두 가지 처리 방법이 있는데, 어떤 부분이 더 낫다고 느끼시나요??

anxi01 replied:
넵 저는 후자를 생각했어요!
required가 false이면 리프레시토큰이 없을 경우 검증 로직이 없었어서 여쭤봤습니다~
아래 방식으로해서 refreshToken 유무 검증 로직만 추가하면 좋을 거 같습니다~

anxi01 commented on src/main/java/spring/backend/auth/application/RotateTokenService.java:
```diff
@@ -0,0 +1,24 @@
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class RotateTokenService {
```
새로운 기능을 하는 서비스와 컨트롤러로 만들어서 하나의 객체의 단일 책임을 갖게 하는 거 아주 좋은 거 같습니다~!

anxi01 commented on src/main/java/spring/backend/auth/application/RotateTokenService.java:
```diff
@@ -0,0 +1,24 @@
+    private final JwtService jwtService;
+
+    public RotateTokenResponse rotateAccessToken(String refreshToken) {
+        UUID memberId = UUID.fromString(jwtService.getPayload(refreshToken).get("memberId", String.class));
```
q) RefreshToken 재발급할 때, 클라이언트에서 받아오는 리프레시 토큰과 레디스에 저장된 리프레시 토큰을 비교해서 같을 경우에만 재발급해줘야하지 않을까요?

anxi01 replied:
q) RefreshToken을 재발급할 때, RefreshToken을 Redis에 갱신해줘야하지 않나요?

HoyeongJeon replied:
엇 이 부분을 체크하지 않았네요! 감사합니다 👍

anxi01 commented on src/main/java/spring/backend/auth/domain/repository/RefreshTokenRepository.java:
```diff
@@ -6,4 +6,5 @@
 public interface RefreshTokenRepository {
     void save(UUID memberId, String refreshToken, Long expireTime, TimeUnit timeUnit);
     String findByMemberId(UUID memberId);
+    void delete(UUID memberId);
```
`deleteByMemberId` 가 명시적으로 의미파악하기 좋을 거 같습니다

anxi01 reviewed (changes requested):
고생 많으셨습니다!
리뷰 확인 부탁드리고, 쿠키로 리프레시 토큰 발급하는거 좋은 거 같습니다!

anxi01 commented on src/main/java/spring/backend/auth/presentation/RotateTokenController.java:
```diff
@@ -0,0 +1,25 @@
+import spring.backend.core.presentation.RestResponse;
+
+@RestController
+@RequestMapping("/v1/token/rotate")
```
q) @RequestMapping에 URI를 넣을 수도 있고, @GetMapping에도 넣을 수 있는데 @GetMapping이 아닌 @RequestMapping에 넣은 이유가 있을까요??

HoyeongJeon replied:
token 을 재발급 해주는 Controller인데, AccessToken 외엔 재발급할 토큰이 없어서 @RequestMapping에 넣었습니다. 

만약 재발급하는 것이 토큰을 제외하고 많았다면

```Java
@RequestMapping("/v1/rotate")
public class RotateTokenController {
  @GetMapping("access_token") 
} 
```
이런식으로 구현했을거 같아요.

다시 보니까 확장성이 고려되지 않은것으로 느껴지네요.

어떤 부분이 더 나아보이시나요?

anxi01 replied:
만약에 각 클래스에 하나의 API를 제공할거라면 
`@RequestMapping("/v1/rotate")` or `@GetMapping("/v1/rotate")`
모두 좋아보입니다~

RefreshToken은 재로그인시에만 재발급되니까 괜찮을 거 같아요!

anxi01 replied:
AccessToken을 재발급해주는 거니까 RotateAccessToken~~ 으로 작성하는게 더 의미파악하기 좋을 거 같아요!

HoyeongJeon committed:
feat: (#35) AuthenticationErrorCode에 쿠키값이 없는 경우 에러를 추가한다.

HoyeongJeon committed:
refactor: (#35) RefreshTokenRedisRepository의 delete 메서드 이름을 deleteByMemberId로 변경한다.

HoyeongJeon committed:
refactor: (#35) RotateTokenController를 RotateAccessTokenController로 변경한다.

HoyeongJeon committed:
feat: (#35) rotateAccessToken 을 수정한다.
- 쿠키에 refreshToken이 존재하지 않은 경우 예외를 발생시킨다.
- 쿠키에 저장된 refreshToken과 저장소에 저장된 refreshToken이 일치하지 않는 경우 예외를 발생시킨다.

HoyeongJeon committed:
refactor: (#35) 서비스에서 처리하던 레포지토리 에러를 레포지토리에 위임한다.

HoyeongJeon committed:
refactor: (#35) RotateTokenService의 이름을 RotateAccessTokenService으로 변경한다.

HoyeongJeon committed:
feat: (#35) RotateAccessTokenService의 테스트 코드를 작성한다.

anxi01 commented on src/main/java/spring/backend/auth/application/RefreshTokenService.java:
```diff
@@ -27,20 +29,33 @@ public RefreshTokenService(JwtService jwtService, @Value("${jwt.refresh-token-ex
+            return refreshToken;
+        } catch (DomainException e) {
+            throw e;
+        } catch (Exception e) {
```
여기서 refreshTokenRepository에서 발생하는 에러는 refreshTokenRepository로 위임하는게 좋아 보입니다.
나중에 refreshTokenRepository 의 메서드를 사용할 때 사용하는 곳(애플리케이션 계층)에서 try-catch를 반복적으로 사용하는 것보다 refreshTokenRepository에서 관련된 에러 처리를 하고 비즈니스 로직에서는 해당 도메인에서 발생하는 에러만 던지면 좋을 거 같습니다.

anxi01 commented on src/main/java/spring/backend/auth/application/RotateAccessTokenService.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    private UUID extractMemberIdFromRefreshToken(String refreshToken) {
+        if(refreshToken == null) {
```
if와 괄호 사이의 공백 추가해주세요~!

anxi01 commented on src/main/java/spring/backend/auth/dto/response/RotateTokenResponse.java:
```diff
@@ -0,0 +1,4 @@
+package spring.backend.auth.dto.response;
+
+public record RotateTokenResponse(String accessToken) {
```
`RotateAccessTokenResponse`가 더 명시적일 거 같습니다!

HoyeongJeon committed:
refactor: (#35) RefreshTokenService의 getRefreshToken에서 발생하는 중 일부를 RefreshTokenRedisRepository로 위임한다.

HoyeongJeon committed:
refactor: (#35) RotateTokenResponse를 RotateAccessTokenResponse로 변경한다.

HoyeongJeon committed:
feat: (#35) RefreshTokenRedisRepositoryTest를 작성하고 RefreshTokenService의 에러를 Repository로 위임한다.

anxi01 reviewed (approved):
고생하셨습니다!
